src/samplers.jl

```diff
@@ -273,3 +274,188 @@ function successive_halving(ho, n, r=1, s=round(Int, log(hb.η, n)))
+end
+
+# sample from KDEMulti
+function sample_potential_hyperparam(kde::MultiKDE.KDEMulti, min_bandwidth, bw_factor)
+    idx = rand(1: size(kde.mat_observations)[2])
+    param = [kde.observations[i][idx] for i in 1:length(kde.observations)]
+    sample = Vector()
+    for (_i, _param, dim_type, _kde) in zip(1:length(kde.dims), param, kde.dims, kde.KDEs)
+        bw = max(_kde.bandwidth, min_bandwidth)
+        local ele
+        if dim_type isa MultiKDE.ContinuousDim
+            bw = bw*bw_factor
+            ele = rand(TruncatedNormal(_param, bw, -_param/bw, (1-_param)/bw))
+        elseif dim_type isa MultiKDE.CategoricalDim
+            if rand() < (1-bw)
+                ele = _param
+            else
+                ele = rand(1:dim_type.levels)
+            end
+        elseif dim_type isa MultiKDE.UnorderedCategoricalDim
+            if rand() < (1-bw)
+                ele = _param
+            else
+                ele = rand(1:dim_type.levels)
+            end
+        else
+            error(string("Dim type ", string(dim_type), " not supported. "))
+        end
+        if kde.mapped[_i]
+            ele = kde.index_to_unordered[_kde][ele]
+        end
+        push!(sample, ele)
+    end
+    sample
+end
+
+# Constructor extensions and adapters for MultiKDE.jl
+const DIMENSION_TYPE = Dict(Categorical=>MultiKDE.CategoricalDim, Continuous=>MultiKDE.ContinuousDim, UnorderedCategorical=>MultiKDE.UnorderedCategoricalDim)
+
+function MultiKDE.KDEMulti(dim_types::Vector{DimensionType}, records::Vector{ObservationsRecord}, min_bandwidth::Real, candidates::Tuple)
+    # Get KDEMulti with min_bandwidth
+    dim = records[1].dim
+    observations = Vector{Vector}()
+    for record in records
+        @assert record.dim == dim "All observations need to be same dimension. "
+        _observations = record.observation
+        if _observations isa Tuple
+            _observations = [_obs for _obs in _observations]
+        end
+        push!(observations, _observations)
+    end
+    # bws = [max(min_bandwidth, KernelDensity.default_bandwidth(observations[i, :])) for i in 1:size(observations)[1]]
+    # candidates = get_dict_candidates(dim_types, candidates)
+    multi_kde = KDEMulti(dim_types, observations, candidates)
+    for i in 1:length(multi_kde.KDEs)
+        if multi_kde.KDEs[i].bandwidth < min_bandwidth
+            multi_kde.KDEs[i].bandwidth = min_bandwidth
+        end
+    end
+    multi_kde
+end
+
+function MultiKDE.KDEMulti(dims::Vector{DimensionType}, observations::Vector, candidates::Tuple)
+    dims = Vector{MultiKDE.DimensionType}([DIMENSION_TYPE[typeof(dim)] === MultiKDE.ContinuousDim ? DIMENSION_TYPE[typeof(dim)]() : 
+                                            DIMENSION_TYPE[typeof(dim)](dim.levels) for dim in dims])
+    MultiKDE.KDEMulti(dims, observations, candidates)
+end
+
+function Base.getproperty(dim_type::Union{MultiKDE.CategoricalDim, MultiKDE.UnorderedCategoricalDim}, v::Symbol)
```

Comment:
What does this method do? It simply calls `getfield` no matter what argument is supplied?